README.md
[![Rust](https://github.com/cjneely10/affinityprop/actions/workflows/rust.yml/badge.svg?branch=main)](https://github.com/cjneely10/affinityprop/actions/workflows/rust.yml)
[![affinityprop crate](https://img.shields.io/crates/v/affinityprop?color=blue&style=flat)](https://crates.io/crates/affinityprop)
[![affinityprop docs](https://img.shields.io/docsrs/affinityprop)](https://docs.rs/affinityprop/0.2.0/affinityprop/)
[![GitHub](https://img.shields.io/github/license/cjneely10/affinityprop)](https://www.gnu.org/licenses/gpl-3.0.html)
[![affinityprop: rustc 1.58](https://img.shields.io/badge/affinityprop-rustc__1.58-blue)](https://doc.rust-lang.org/rustc/what-is-rustc.html)
![coverage](https://img.shields.io/badge/coverage-95%25-success)

# affinityprop
The `affinityprop` crate provides an optimized implementation of the Affinity Propagation
clustering algorithm, which identifies cluster of data without *a priori* knowledge about the
number of clusters in the data. The original algorithm was developed by
<a href="http://utstat.toronto.edu/reid/sta414/frey-affinity.pdf" target="_blank">Brendan Frey and Delbery Dueck</a>

# About
Affinity Propagation identifies a subset of representative examples from a dataset, known as
**exemplars**.

Briefly, the algorithm accepts as input a matrix describing **pairwise similarity** for all data
values. This information is used to calculate pairwise **responsibility** and **availability**.
Responsibility *`r(i,j)`* describes how well-suited point *`j`* is to act as an exemplar for
point *`i`* when compared to other potential exemplars. Availability *`a(i,j)`* describes how
appropriate it is for point *i* to accept point *j* as its exemplar when compared to
other exemplars.

Users provide a number of **convergence iterations** to repeat the calculations, after which the
potential exemplars are extracted from the dataset. Then, the algorithm continues to repeat
until the exemplar values stop changing, or until the **maximum iterations** are met.

# Why this crate?
The nature of Affinity Propagation demands an *O(n<sup>2</sup>)* runtime. An existing
<a href="https://github.com/scikit-learn/scikit-learn/blob/0d378913b/sklearn/cluster/_affinity_propagation.py#L38" target="_blank">sklearn</a>
version is implemented using the Python library
<a href="https://numpy.org/doc/stable/index.html" target="_blank">numpy</a>
which incorporates vectorized row operations. Coupled with **SIMD** instructions, this results
in decreased time to finish.

However, in applications with large input values, the *O(n<sup>2</sup>)* runtime is still
prohibitive. This crate implements Affinity Propagation using the
<a href="https://crates.io/crates/rayon" target="_blank">rayon</a>
crate, which allows for a drastic decrease in overall runtime - as much as 30-60% when compiled
in release mode!

# Dependencies
<a href="https://doc.rust-lang.org/cargo/getting-started/installation.html">cargo</a>
with `rustc >=1.58`

# Installation
## In Rust code
Latest version:

```toml
[dependencies]
affinityprop = "https://github.com/cjneely10/affinityprop"
ndarray = "0.15.4"
```

Current stable version:

```toml
[dependencies]
affinityprop = "0.2.0"
ndarray = "0.15.4"
```

## As a command-line tool
Latest version:

```shell
cargo install --git https://github.com/cjneely10/affinityprop
```

Current stable version:

```shell
cargo install affinityprop
```

# Usage

## From Rust code
The `affinityprop` crate expects a type that defines how to calculate pairwise `Similarity`
for all data points. This crate provides the `NegEuclidean`, `NegCosine`, and
`LogEuclidean` structs, which are defined as `-1 * sum((a - b)**2)`, `-1 * (a . b)/(|a|*|b|)`,
and `sum(log((a - b)**2))`, respectively.

Users who wish to calculate similarity differently are advised that **Affinity Propagation
expects *s(i,j)* > *s(i, k)* iff *i* is more similar to *j* than it is to *k***.

```rust
use ndarray::{arr1, arr2, Array2};
use affinityprop::{AffinityPropagation, NegCosine, Preference};

let x: Array2<f32> = arr2(&[[0., 1., 0.], [2., 3., 2.], [3., 2., 3.]]);

// Cluster using negative cosine similarity with a pre-defined preference
let ap = AffinityPropagation::default();
let (converged, results) = ap.predict(&x, NegCosine::default(), Preference::Value(-10.));
assert!(converged && results.len() == 1 && results.contains_key(&0));

// Cluster with list of preference values
let pref = arr1(&[0., -1., 0.]);
let (converged, results) = ap.predict(&x, NegCosine::default(), Preference::List(&pref));
assert!(converged);
assert!(results.len() == 2 && results.contains_key(&0) && results.contains_key(&2));

// Use damping=0.5, threads=2, convergence_iter=10, max_iterations=100,
// median similarity as preference
let ap = AffinityPropagation::new(0.5, 2, 10, 100);
let (converged, results) = ap.predict(&x, NegCosine::default(), Preference::Median);
assert!(converged);
assert!(results.len() == 2 && results.contains_key(&0) && results.contains_key(&2));

// Predict with pre-calculated similarity
let s: Array2<f32> = arr2(&[[0., -3., -12.], [-3., 0., -3.], [-12., -3., 0.]]);
let ap = AffinityPropagation::default();
let (converged, results) = ap.predict_precalculated(s, Preference::Value(-10.));
assert!(converged && results.len() == 1 && results.contains_key(&1));
```

**Note on `NaN`**: `f32::NAN` and `f64::NAN` are allowed as input into `AffinityPropagation::predict`, but will cause the program
to panic if provided in `AffinityPropagation::predict_precalculated`.

## From the Command Line
`affinityprop` can be run from the command-line and used to analyze a file of data:

```text
ID1 val1 val2
ID2 val3 val4
ID3 val5 val6
```

where ID*n* is any string identifier and val*n* are floating-point (decimal) values.

`nan` values are allowed as input when using a provided similarity calculation, but NOT for `-s 3` precalculated similarity input matrices.
The program will notify the user of `nan` input, but will only terminate if an entire data row of `nan` values is encountered.

The file delimiter is provided from the command line with the `-l` flag.
Similarity will be calculated based on the option set by the `-s` flag.

For files without row ids:

```text
val1 val2
val3 val4
val5 val6
```

provide the `-n` flag from the command line. IDs will automatically be assigned by zero-based
index.

Users may instead provide a pre-calculated similarity matrix by passing the `-s 3` flag from
the command line and by structuring their input file as:

```text
ID1 sim11 sim12 sim13
ID2 sim21 sim22 sim23
ID3 sim31 sim32 sim33
```

where row*i*, col*j* is the pairwise similarity between inputs *i* and *j*.

Or, for files without row labels, users may pass `-n -s 3`:

```text
sim11 sim12 sim13
sim21 sim22 sim23
sim31 sim32 sim33
```

IDs will automatically be assigned by zero-based index.

### Help Menu
```text
affinityprop 0.3.0-alpha
Chris N. <christopher.neely1200@gmail.com>
Vectorized and Parallelized Affinity Propagation

USAGE:
    affinityprop [OPTIONS] --input <INPUT>

FLAGS:
    -n, --no_labels    Input file does not contain IDS as the first column
    -h, --help         Prints help information
    -V, --version      Prints version information

OPTIONS:
    -c, --convergence_iter <CONV_ITER>    Convergence iterations, default=10
    -d, --damping <DAMPING>               Damping value in range (0, 1), default=0.9
    -l, --delimiter <DELIMITER>           File delimiter, default '\t'
    -i, --input <INPUT>                   Path to input file
    -m, --max_iter <MAX_ITER>             Maximum iterations, default=100
    -r, --precision <PRECISION>           Set f32 or f64 precision, default=f32
    -p, --preference <PREF>               Preference to be own exemplar, default=median pairwise similarity
    -s, --similarity <SIMILARITY>         Set similarity calculation method
                                          (0=NegEuclidean,1=NegCosine,2=LogEuclidean,3=precalculated), default=0
    -t, --threads <THREADS>               Number of worker threads, default=4
```

### Results
Results are printed to stdout in the format:

```text
Converged=true/false nClusters=NC nSamples=NS
>Cluster=n size=N exemplar=i
[comma-separated cluster member IDs/indices]
>Cluster=n size=N exemplar=i
[comma-separated cluster member IDs/indices]
...
```

# Runtime and Resource Notes
Affinity Propagation is *O(n<sup>2</sup>)* in both runtime and memory.
This crate seeks to address the former, not the latter.

An estimated memory usage can be calculated given:

```text
memory(GB) = p * 4 * N^2 / 2^30
```

For `N` inputs. `p = 4` for 32-bit floating-point precision and `p = 8` for 64-bit.

## Comparison to sklearn implementation

This implementation was tested on 50-D isotropic Gaussian blobs generated by the sklearn 
[make_blobs](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_blobs.html)
function (n=300,10000,25000). A [dataset of biological relevance](https://github.com/edgraham/BinSanity/blob/master/example/Infant_gut_assembly.cov.x100.lognorm)
was also selected (n=4189). ARI/F1 scores >= 0.99 were obtained for the Gaussian data when compared to the labels 
provided by `make_blobs`. ARI/F1 = 0.98 was obtained for the biological dataset when compared to `sklearn`-derived 
labels.

This `affinityprop` implementation was compared against the [Affinity Propagation](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AffinityPropagation.html#sklearn.cluster.AffinityPropagation.fit_predict) 
implementation contained within scikit-learn-1.0.2, and run using numpy-1.22.2 with Python 3.10.0. 
This analysis was completed using a Ryzen 9 3950X processor. 

In all analyses, damping=0.95, convergence_iter=400, and max_iter=4000. 
Preference=-1000.0 for Gaussian data and -10.0 for biological data.

![Time to complete analysis, scaled by the sklearn implementation, plotted against the number of cores.](assets/combined-data.png)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]

### Python bindings
- [PyO3](https://github.com/PyO3/pyo3)-derived bindings for use within Python

## [0.3.0-alpha] - 2022-07-01
### Changed
- NaN values are supported in input matrices
  - Previously, cli would exit on nan detection
  - Now, notify user via stderr
    - Only exit if full row of NaN found
  - Provided similarity calculations negate NaN effectd by replacing them in final sum operations with `0`
    - See `NegEuclidean`'s `Similarity` implementation
- NaN check in `predict_precalculated`
  - Panic on detection

## [0.2.0] - 2022-06-24
### Added
- Better stderr messages on file-parsing errors
  - Line/col numbers and kind of error

### Changed
- Small edits to pass-by-value instead of by-reference

### Fixed
- `nan` values were accepted as valid input.
  - No valid similarity calculation is present for `nan` values

## [0.1.1] - 2022-02-15
First release of `affinityprop`


tests/normal.rs
use std::collections::HashMap;
use std::ffi::OsStr;
use std::fmt::Debug;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::{Path, PathBuf};
use std::str::FromStr;

use ndarray::{Array2, Axis};
use num_traits::Float;

use affinityprop::{AffinityPropagation, NegEuclidean};

fn load_data<F>(test_file: PathBuf) -> std::io::Result<(Array2<F>, HashMap<usize, Vec<usize>>)>
where
    F: Float + Send + Sync + FromStr + Default,
    <F as FromStr>::Err: Debug,
{
    let reader = BufReader::new(File::open(test_file)?);
    let mut test_data_map: HashMap<usize, Vec<usize>> = HashMap::new();
    let mut test_data = Vec::new();
    reader
        .lines()
        .map(|l| l.unwrap())
        .enumerate()
        .for_each(|(idx, line)| {
            let mut line = line.split(" ");
            let id = line.next().unwrap().parse::<usize>().unwrap();
            if !test_data_map.contains_key(&id) {
                test_data_map.insert(id, vec![]);
            }
            test_data_map.get_mut(&id).unwrap().push(idx);
            let point: Vec<F> = line.map(|c| c.parse::<F>().unwrap()).collect();
            test_data.push(point);
        });
    let mut out = Array2::<F>::default((test_data.len(), test_data[0].len()));
    out.axis_iter_mut(Axis(0))
        .enumerate()
        .for_each(|(idx1, mut row)| {
            row.iter_mut().enumerate().for_each(|(idx2, col)| {
                *col = test_data[idx1][idx2];
            });
        });
    Ok((out, test_data_map))
}

fn run_test(path: PathBuf) {
    let (test_array, test_results) = load_data::<f32>(path).unwrap();
    let ap = AffinityPropagation::new(-1000., 0.5, 4, 400, 4000);
    let (converged, results) = ap.predict(&test_array, NegEuclidean::default());
    assert!(converged);
    assert_eq!(test_results.len(), results.len());
}

fn file<A: AsRef<OsStr>>(path: A) -> PathBuf {
    let test_dir = Path::new(file!()).parent().unwrap();
    test_dir.join(Path::new("data")).join(Path::new(&path))
}

#[test]
fn ten_exemplars() {
    run_test(file(&"near-exemplar-10.test"));
}

#[test]
fn one_hundred_exemplars() {
    run_test(file(&"near-exemplar-100.test"));
}


tests/simple.rs
#[cfg(test)]
mod test {
    use ndarray::{arr2, Array2};

    use affinityprop::{AffinityPropagation, Euclidean};

    #[test]
    fn simple() {
        let x: Array2<f32> = arr2(&[[1., 1., 1.], [2., 2., 2.], [3., 3., 3.]]);
        let y = vec!["1", "2", "3"];
        let mut ap = AffinityPropagation::default();
        let results = ap.predict(x, &y, Euclidean::default());
        assert!(results.len() == 1 && results.contains_key(&"2"));
    }
}


tests/generate_ig.py
#!/usr/bin/env python
"""
Generate `binsanity.test` file from BinSanity example data.

Expects env variable set that points to directory of results from
running BinSanity example

Tested using Python 3.6.13


Usage:
./generate_ig
"""
import os
from pathlib import Path

# Directory of results from BinSanity example
from typing import Dict

FASTA_RESULTS_DIR_A = os.environ.get("BINSANITY_RESULTS_DIR_A")
if FASTA_RESULTS_DIR_A is None:
    exit("BINSANITY_RESULTS_DIR_A is not set")

FASTA_RESULTS_DIR_B = os.environ.get("BINSANITY_RESULTS_DIR_B")
if FASTA_RESULTS_DIR_B is None:
    exit("BINSANITY_RESULTS_DIR_B is not set")

# Write directory
DATA_DIR = "test-data"


def get_ids_from_fasta(fasta_file: Path) -> set:
    ids = set()
    with open(fasta_file, "r") as file_ptr:
        for line in file_ptr:
            if line.startswith(">"):
                ids.add(line.split(" ")[0][1:].rstrip("\r\n"))
    return ids


def get_data_from_ig_file(ig_file: Path) -> Dict[str, str]:
    _point_data = {}
    with open(ig_file, "r") as ig_ptr:
        for line in ig_ptr:
            data = line.split("\t")
            _point_data[data[0]] = " ".join(data[1:])
    return _point_data


if __name__ == "__main__":
    with open(f"{DATA_DIR}/binsanity.test", "w") as out_ptr:
        cluster_id = 0
        point_data = get_data_from_ig_file(Path(DATA_DIR).joinpath("Infant_gut_assembly.cov.x100.lognorm"))
        for file in os.listdir(FASTA_RESULTS_DIR_A):
            if file.endswith(".fna"):
                file = Path(os.path.join(FASTA_RESULTS_DIR_A, file)).resolve()
                for _id in get_ids_from_fasta(file):
                    out_ptr.write(f"{cluster_id} {point_data[_id]}")
                cluster_id += 1
    with open(f"{DATA_DIR}/binsanity.2.test", "w") as out_ptr:
        cluster_id = 0
        point_data = get_data_from_ig_file(Path(DATA_DIR).joinpath("assembly.log.cov.x100.lognorm"))
        for file in os.listdir(FASTA_RESULTS_DIR_B):
            if file.endswith(".fna"):
                file = Path(os.path.join(FASTA_RESULTS_DIR_B, file)).resolve()
                for _id in get_ids_from_fasta(file):
                    out_ptr.write(f"{cluster_id} {point_data[_id]}")
                cluster_id += 1


src/ops.rs
use crate::affinity_propagation::Value;
use ndarray::{Array2, Axis};
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::PathBuf;

/// Reads in a file formatted as (tab separated):
///     id1 val1 val2 val3
///     id2 val1 val2 val3
///
/// Provide as many ids and values as desired
/// All rows should be same length
/// Values should be floating-point decimal values
pub(crate) fn from_file(p: PathBuf) -> (Array2<Value>, Vec<String>) {
    let reader = BufReader::new(File::open(p).unwrap());
    let mut labels = Vec::new();
    let mut data = Vec::new();
    // Read tab-delimited file
    reader.lines().map(|l| l.unwrap()).for_each(|line| {
        let mut line = line.split('\t');
        // ID as first col
        labels.push(line.next().expect("Error loading line label").to_string());
        // Rest are data
        data.push(
            line.map(|s| s.parse::<Value>().expect("Error parsing data in file"))
                .collect::<Vec<Value>>(),
        );
    });
    // Validate data was loaded
    assert!(
        data.len() > 1,
        "Data file is empty or only contains a single entry"
    );
    // Validate data all has same length
    let length = data[0].len();
    data.iter().skip(1).for_each(|v| {
        assert_eq!(v.len(), length, "Input data rows must all be same length!");
    });
    // Convert data to Array2
    let mut out = Array2::<Value>::default((data.len(), data[0].len()));
    out.axis_iter_mut(Axis(0))
        .enumerate()
        .for_each(|(idx1, mut row)| {
            row.iter_mut().enumerate().for_each(|(idx2, col)| {
                *col = data[idx1][idx2];
            });
        });
    (out, labels)
}


src/main.rs
use crate::affinity_propagation::{AffinityPropagation, Config, Euclidean, Value};
use crate::ops::from_file;
use std::path::Path;
use std::process::exit;

mod affinity_propagation;
mod ops;

#[macro_use]
extern crate clap;

fn main() {
    let matches = clap_app!(affinityprop =>
        (version: "0.1.0")
        (author: "Chris N. <christopher.neely1200@gmail.com>")
        (about: "Vectorized and Parallelized Affinity Propagation")
        (@arg INPUT: -i --input +takes_value +required "Path to input file")
        (@arg PREF: -p --preference +takes_value +allow_hyphen_values "Preference, default=-10.0")
        (@arg MAX_ITER: -m --max_iter +takes_value "Maximum iterations, default=100")
        (@arg CONV_ITER: -c --convergence_iter +takes_value "Convergence iterations, default=10")
        (@arg DAMPING: -d --damping +takes_value "Damping value, default=0.9")
        (@arg THREADS: -t --threads +takes_value "Number of worker threads, default=4")
    )
    .get_matches();

    let input_file = matches.value_of("INPUT").unwrap().to_string();
    if !Path::new(&input_file).exists() {
        eprintln!("Unable to locate input file {}", input_file);
        exit(1);
    }
    let preference = matches
        .value_of("PREF")
        .unwrap_or("-10.0")
        .parse::<Value>()
        .expect("Unable to parse preference");
    let mut max_iterations = matches
        .value_of("MAX_ITER")
        .unwrap_or("100")
        .parse::<usize>()
        .expect("Unable to parse max_iterations");
    let convergence_iter = matches
        .value_of("CONV_ITER")
        .unwrap_or("10")
        .parse::<usize>()
        .expect("Unable to parse convergence_iter");
    let damping = matches
        .value_of("DAMPING")
        .unwrap_or("0.9")
        .parse::<Value>()
        .expect("Unable to parse damping");
    let threads = matches
        .value_of("THREADS")
        .unwrap_or("4")
        .parse::<usize>()
        .expect("Unable to parse threads");
    // Validate values
    if threads < 1 || (damping < 0. || damping > 1.) || convergence_iter < 1 || max_iterations < 1 {
        eprintln!("Improper parameter set!");
        exit(2);
    }
    if convergence_iter > max_iterations {
        max_iterations = convergence_iter * 2;
    }
    let (x, y) = from_file(Path::new(&input_file).to_path_buf());

    // Run AP
    let cfg = Config {
        preference,
        max_iterations,
        convergence_iter,
        damping,
        threads,
    };
    println!("{:?}", cfg);
    AffinityPropagation::predict(x, y, cfg, Euclidean::default());
}


src/array.rs
#[derive(Debug)]
pub struct Array<T> {
    vec: Vec<T>,
    row_size: usize,
}

impl<T> Array<T>
where
    T: Default + Copy,
{
    pub fn new(r: usize, c: usize) -> Self {
        Self {
            vec: vec![T::default(); r * c],
            row_size: r,
        }
    }

    pub fn rows(&self) -> usize {
        self.row_size
    }

    /// Get value at index in array
    pub fn at(&self, i: usize) -> Option<&T> {
        self.vec.get(i)
    }

    /// Get value at (r,c) index in array
    pub fn get(&self, r: usize, c: usize) -> Option<&T> {
        self.vec.get(r * self.row_size + c)
    }

    pub fn set(&mut self, r: usize, c: usize, val: T) {
        self.vec[r * self.row_size + c] = val;
    }
}

impl<T> From<&Array<T>> for Array<T>
where
    T: Default + Copy,
{
    fn from(f: &Array<T>) -> Self {
        Self {
            vec: f.vec.clone(),
            row_size: f.row_size,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple() {
        let mut array = Array::new(1000, 1000);
        let mut counter = 0;
        (0..1000)
            .into_iter()
            .map(|a| {
                (0..1000)
                    .into_iter()
                    .map(|b| {
                        array.set(a, b, counter);
                        counter += 1;
                    })
                    .last();
            })
            .last();
        (0..1000)
            .into_iter()
            .rev()
            .map(|a| {
                (0..1000)
                    .into_iter()
                    .rev()
                    .map(|b| {
                        let val = array.get(a, b);
                        assert_eq!(val.unwrap(), &(counter - 1));
                        counter -= 1;
                    })
                    .last();
            })
            .last();
    }

    #[test]
    #[should_panic]
    fn oob() {
        let mut array: Array<u8> = Array::new(10, 10);
        assert_eq!(None, array.get(11, 9));
        array.set(11, 9, 10);
    }

    #[test]
    fn at_get() {
        let mut array: Array<u8> = Array::new(10, 10);
        array.set(5, 5, 1);
        assert_eq!(array.get(5, 5), array.at(5 * 10 + 5));
    }
}


src/preference.rs
use ndarray::{Array1, Array2, Zip};
use num_traits::Float;

/// Preference is the value representing the degree to which a data point will act as its own exemplar,
/// with lower (more negative) values yielding fewer clusters.
///
/// - Median: Use median similarity value as preference
/// - List: Use provided preference list
/// - Value: Assign all members the same preference value
#[derive(Debug, Clone)]
pub enum Preference<'a, F>
where
    F: Float + Send + Sync,
{
    /// Use the median pairwise similarity as the preference
    Median,
    /// Use a list of preferences, one per input
    List(&'a Array1<F>),
    /// Use a single value as the preference for all inputs
    Value(F),
}

pub(crate) fn place_preference<F>(s: &mut Array2<F>, p: Preference<F>)
where
    F: Float + Send + Sync,
{
    let s_dim = s.dim();
    let preference = match p {
        Preference::Value(pref) => pref,
        Preference::Median => median(s),
        Preference::List(l) => {
            assert_eq!(
                s_dim.0,
                l.len(),
                "Preference list length does not match input length!"
            );
            Zip::from(l)
                .and(s.diag_mut())
                .par_for_each(|pref, s_pos| *s_pos = *pref);
            return;
        }
    };
    s.diag_mut().par_map_inplace(|v| *v = preference);
}

/// Computed simply - collect values into vector, sort, and return value at len() / 2
fn median<F>(x: &Array2<F>) -> F
where
    F: Float + Send + Sync,
{
    let mut sorted_values = Vec::new();
    let x_dim_0 = x.dim().0 as usize;
    for i in 0..x_dim_0 {
        for j in (i + 1)..x_dim_0 {
            sorted_values.push(x[[i, j]]);
        }
    }
    sorted_values.sort_by(|a, b| a.partial_cmp(b).unwrap());
    sorted_values[sorted_values.len() / 2]
}

#[cfg(test)]
mod test {
    use ndarray::{arr1, arr2, Array2};
    use rayon::ThreadPool;

    use crate::preference::{median, place_preference};
    use crate::Preference::List;

    fn pool(t: usize) -> ThreadPool {
        rayon::ThreadPoolBuilder::new()
            .num_threads(t)
            .build()
            .unwrap()
    }

    fn test_data() -> Array2<f32> {
        arr2(&[
            [0., -5., -6., -12., -17.],
            [-5., 0., -17., -17., -22.],
            [-6., -17., 0., -18., -21.],
            [-12., -17., -18., 0., -3.],
            [-17., -22., -21., -3., 0.],
        ])
    }

    #[test]
    fn valid_median() {
        assert_eq!(-17., median(&test_data()));
    }

    #[test]
    fn provided_preference_list() {
        pool(2).scope(move |_| {
            let mut sim = test_data();
            let pref_list = arr1(&[-1., -2., -3., -4., -5.]);
            place_preference(&mut sim, List(&pref_list));
        });
    }

    #[test]
    #[should_panic]
    fn invalid_preference_list() {
        pool(2).scope(move |_| {
            let mut sim = test_data();
            let pref_list = arr1(&[-1., -2., -3.]);
            place_preference(&mut sim, List(&pref_list));
        });
    }
}


src/lib.rs
//! The `affinityprop` crate provides an optimized implementation of the Affinity Propagation
//! clustering algorithm, which identifies cluster of data without *a priori* knowledge about the
//! number of clusters in the data. The original algorithm was developed by
//! <a href="http://utstat.toronto.edu/reid/sta414/frey-affinity.pdf" target="_blank">Brendan Frey and Delbery Dueck</a>
//!
//! # About
//! Affinity Propagation identifies a subset of representative examples from a dataset, known as
//! **exemplars**.
//!
//! Briefly, the algorithm accepts as input a matrix describing **pairwise similarity** for all data
//! values. This information is used to calculate pairwise **responsibility** and **availability**.
//! Responsibility *`r(i,j)`* describes how well-suited point *`j`* is to act as an exemplar for
//! point *`i`* when compared to other potential exemplars. Availability *`a(i,j)`* describes how
//! appropriate it is for point *i* to accept point *j* as its exemplar when compared to
//! other exemplars.
//!
//! Users provide a number of **convergence iterations** to repeat the calculations, after which the
//! potential exemplars are extracted from the dataset. Then, the algorithm continues to repeat
//! until the exemplar values stop changing, or until the **maximum iterations** are met.
//!
//! # Why this crate?
//! The nature of Affinity Propagation demands an *O(n<sup>2</sup>)* runtime. An existing
//! <a href="https://github.com/scikit-learn/scikit-learn/blob/0d378913b/sklearn/cluster/_affinity_propagation.py#L38" target="_blank">sklearn</a>
//! version is implemented using the Python library
//! <a href="https://numpy.org/doc/stable/index.html" target="_blank">numpy</a>
//! which incorporates vectorized row operations. Coupled with **SIMD** instructions, this results
//! in decreased time to finish.
//!
//! However, in applications with large input values, the *O(n<sup>2</sup>)* runtime is still
//! prohibitive. This crate implements Affinity Propagation using the
//! <a href="https://crates.io/crates/rayon" target="_blank">rayon</a>
//! crate, which allows for a drastic decrease in overall runtime - as much as 30-60% when compiled
//! in release mode!
//!
//! # Dependencies
//! <a href="https://doc.rust-lang.org/cargo/getting-started/installation.html">cargo</a>
//! with `rustc >=1.58`
//!
//! # Installation
//! ## In Rust code
//! ```toml
//! [dependencies]
//! affinityprop = "0.2.0"
//! ndarray = "0.15.4"
//! ```
//!
//! ## As a command-line tool
//! ```shell
//! cargo install affinityprop
//! ```
//!
//! # Usage
//!
//! ## From Rust code
//! The `affinityprop` crate expects a type that defines how to calculate pairwise [`Similarity`]
//! for all data points. This crate provides the [`NegEuclidean`], [`NegCosine`], and
//! [`LogEuclidean`] structs, which are defined as `-1 * sum((a - b)**2)`, `-1 * (a . b)/(|a|*|b|)`,
//! and `sum(log((a - b)**2))`, respectively.
//!
//! Users who wish to calculate similarity differently are advised that **Affinity Propagation
//! expects *s(i,j)* > *s(i, k)* iff *i* is more similar to *j* than it is to *k***.
//!
//!     use ndarray::{arr1, arr2, Array2};
//!     use affinityprop::{AffinityPropagation, NegCosine, Preference};
//!     
//!     let x: Array2<f32> = arr2(&[[0., 1., 0.], [2., 3., 2.], [3., 2., 3.]]);
//!
//!     // Cluster using negative cosine similarity with a pre-defined preference
//!     let ap = AffinityPropagation::default();
//!     let (converged, results) = ap.predict(&x, NegCosine::default(), Preference::Value(-10.));
//!     assert!(converged && results.len() == 1 && results.contains_key(&0));
//!
//!     // Cluster with list of preference values
//!     let pref = arr1(&[0., -1., 0.]);
//!     let (converged, results) = ap.predict(&x, NegCosine::default(), Preference::List(&pref));
//!     assert!(converged);
//!     assert!(results.len() == 2 && results.contains_key(&0) && results.contains_key(&2));
//!
//!     // Use damping=0.5, threads=2, convergence_iter=10, max_iterations=100,
//!     // median similarity as preference
//!     let ap = AffinityPropagation::new(0.5, 2, 10, 100);
//!     let (converged, results) = ap.predict(&x, NegCosine::default(), Preference::Median);
//!     assert!(converged);
//!     assert!(results.len() == 2 && results.contains_key(&0) && results.contains_key(&2));
//!
//!     // Predict with pre-calculated similarity
//!     let s: Array2<f32> = arr2(&[[0., -3., -12.], [-3., 0., -3.], [-12., -3., 0.]]);
//!     let ap = AffinityPropagation::default();
//!     let (converged, results) = ap.predict_precalculated(s, Preference::Value(-10.));
//!     assert!(converged && results.len() == 1 && results.contains_key(&1));
//!
//! ## From the Command Line
//! `affinityprop` can be run from the command-line and used to analyze a file of data:
//!
//! ```text
//! ID1 val1 val2
//! ID2 val3 val4
//! ID3 val5 val6
//! ```
//!
//! where ID*n* is any string identifier and val*n* are floating-point (decimal) values.
//! The file delimiter is provided from the command line with the `-l` flag.
//! Similarity will be calculated based on the option set by the `-s` flag.
//!
//! For files without row ids:
//!
//! ```text
//! val1 val2
//! val3 val4
//! val5 val6
//! ```
//!
//! provide the `-n` flag from the command line. IDs will automatically be assigned by zero-based
//! index.
//!
//! Users may instead provide a pre-calculated similarity matrix by passing the `-s 3` flag from
//! the command line and by structuring their input file as:
//!
//! ```text
//! ID1 sim11 sim12 sim13
//! ID2 sim21 sim22 sim23
//! ID3 sim31 sim32 sim33
//! ```
//!
//! where row*i*, col*j* is the pairwise similarity between inputs *i* and *j*.
//!
//! Or, for files without row labels, users may pass `-n -s 3`:
//!
//! ```text
//! sim11 sim12 sim13
//! sim21 sim22 sim23
//! sim31 sim32 sim33
//! ```
//!
//! IDs will automatically be assigned by zero-based index.
//!
//! ### Help Menu
//! ```text
//! affinityprop 0.3.0-alpha
//! Chris N. <christopher.neely1200@gmail.com>
//! Vectorized and Parallelized Affinity Propagation
//!
//! USAGE:
//!     affinityprop [OPTIONS] --input <INPUT>
//!
//! FLAGS:
//!     -n, --no_labels    Input file does not contain IDS as the first column
//!     -h, --help         Prints help information
//!     -V, --version      Prints version information
//!
//! OPTIONS:
//!     -c, --convergence_iter <CONV_ITER>    Convergence iterations, default=10
//!     -d, --damping <DAMPING>               Damping value in range (0, 1), default=0.9
//!     -l, --delimiter <DELIMITER>           File delimiter, default '\t'
//!     -i, --input <INPUT>                   Path to input file
//!     -m, --max_iter <MAX_ITER>             Maximum iterations, default=100
//!     -r, --precision <PRECISION>           Set f32 or f64 precision, default=f32
//!     -p, --preference <PREF>               Preference to be own exemplar, default=median pairwise similarity
//!     -s, --similarity <SIMILARITY>         Set similarity calculation method
//!                                           (0=NegEuclidean,1=NegCosine,2=LogEuclidean,3=precalculated), default=0
//!     -t, --threads <THREADS>               Number of worker threads, default=4
//! ```
//!
//! ### Results
//! Results are printed to stdout in the format:
//!
//! ```text
//! Converged=true/false nClusters=NC nSamples=NS
//! >Cluster=n size=N exemplar=i
//! [comma-separated cluster member IDs/indices]
//! >Cluster=n size=N exemplar=i
//! [comma-separated cluster member IDs/indices]
//! ...
//! ```
//!
//! # Runtime and Resource Notes
//! Affinity Propagation is *O(n<sup>2</sup>)* in both runtime and memory.
//! This crate seeks to address the former, not the latter.
//!
//!
//! An estimated memory usage can be calculated given:
//!
//! ```text
//! memory(GB) = p * 4 * N^2 / 2^30
//! ```
//!
//! For `N` inputs. `p = 4` for 32-bit floating-point precision and `p = 8` for 64-bit.
//!
pub use affinity_propagation::{AffinityPropagation, Cluster, ClusterMap, ClusterResults, Idx};
pub use preference::Preference;
pub use similarity::{LogEuclidean, NegCosine, NegEuclidean, Similarity};

mod affinity_propagation;
mod algorithm;
mod preference;
mod similarity;
